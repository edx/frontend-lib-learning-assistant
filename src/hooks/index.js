export { default as useCourseUpgrade } from './use-course-upgrade';
export { default as useTrackEvent } from './use-track-event';
