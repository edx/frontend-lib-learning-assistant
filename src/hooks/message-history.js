/* eslint-disable import/prefer-default-export */
import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { getLearningAssistantMessageHistory } from '../data/thunks';

export const useMessageHistory = (courseId) => {
  const dispatch = useDispatch();
  const { isEnabled } = useSelector(state => state.learningAssistant);

  useEffect(() => {
    if (!courseId || !isEnabled) { return; }

    dispatch(getLearningAssistantMessageHistory(courseId));
  }, [dispatch, isEnabled, courseId]);
};
