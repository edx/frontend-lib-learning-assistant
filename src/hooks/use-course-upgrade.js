import { useContext } from 'react';
import { useModel } from '@src/generic/model-store'; // eslint-disable-line import/no-unresolved
import { useSelector } from 'react-redux';
import { CourseInfoContext } from '../context';

const millisecondsInOneDay = 24 * 60 * 60 * 1000; // hours*minutes*seconds*milliseconds

/**
 * @typedef AuditTrial
 * @type {object}
 * @property {string} startDate Date String of when the Trial started.
 * @property {string} expirationDate Date String of when the Trial ends.
 */

/**
 * @typedef CourseUpgradeInfo
 * @type {object}
 * @property {boolean} upgradeable Should this user see a trial/upgrade option?.
 * @property {number} [auditTrialLengthDays] Audit Trial full length in days.
 * @property {number} [auditTrialDaysRemaining] Remaining day for the current trial (if any).
 * @property {boolean} [auditTrialExpired] True means that the audit has been taken and expired.
 * @property {AuditTrial} [auditTrial] The Audit trial information. Undefined if there's no trial for this user.
 * @property {string} [upgradeUrl] URL to redirect the user in case there's intention to upgrade.
 */

/**
 * This hook returns related data for the Course Upgrade.
 *
 * @returns {CourseUpgradeInfo}
 */
export default function useCourseUpgrade() {
  const { courseId, isUpgradeEligible } = useContext(CourseInfoContext);
  const {
    offer,
    accessExpiration,
    datesBannerInfo,
  } = useModel('coursewareMeta', courseId);
  const { verifiedMode } = useModel('courseHomeMeta', courseId);
  const {
    auditTrialLengthDays,
    auditTrial,
  } = useSelector(state => state.learningAssistant);

  const upgradeUrl = offer?.upgradeUrl || verifiedMode?.upgradeUrl;

  if (!isUpgradeEligible || !upgradeUrl) { return { upgradeable: false }; }

  let auditTrialExpired = false;
  let auditTrialDaysRemaining;

  if (auditTrial?.expirationDate) {
    const auditTrialExpirationDate = new Date(auditTrial.expirationDate);
    auditTrialDaysRemaining = Math.ceil((auditTrialExpirationDate - Date.now()) / millisecondsInOneDay);

    auditTrialExpired = auditTrialDaysRemaining < 0;
  }

  const isFBE = !!accessExpiration && !!datesBannerInfo?.contentTypeGatingEnabled;

  return {
    upgradeable: true,
    auditTrialLengthDays,
    auditTrialDaysRemaining,
    auditTrialExpired,
    auditTrial,
    upgradeUrl,
    isFBE,
  };
}
