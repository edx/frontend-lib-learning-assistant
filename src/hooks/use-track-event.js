import { useContext } from 'react';
import { useModel } from '@src/generic/model-store'; // eslint-disable-line import/no-unresolved
import { getAuthenticatedUser } from '@edx/frontend-platform/auth';
import { sendTrackEvent } from '@edx/frontend-platform/analytics';
import { CourseInfoContext } from '../context';

/**
 * @typedef Tracker
 * @type {object}
 * @property {(eventName: string, details?: object)} track Calls sendTrackEvent with user and course context.
 */

/**
 * This hook returns a track method to track events.
 *
 * @returns {Tracker}
 */
export default function useTrackEvent() {
  const { courseId, moduleId } = useContext(CourseInfoContext);
  const { org } = useModel('courseHomeMeta', courseId);
  const { userId } = getAuthenticatedUser();

  const track = (event, details) => {
    sendTrackEvent(event, {
      course_id: courseId,
      org_key: org,
      user_id: userId,
      module_id: moduleId,
      ...details,
    });
  };

  return { track };
}
