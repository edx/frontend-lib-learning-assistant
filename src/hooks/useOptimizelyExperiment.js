import { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { useDecision } from '@optimizely/react-sdk';
import { getAuthenticatedUser } from '@edx/frontend-platform/auth';
import { setExperiment } from '../data/slice';

// We need this import to make sure Optimizely is instantiated.
import optimizelyInstance from '../data/optimizely'; // eslint-disable-line no-unused-vars

const useOptimizelyExperiment = (flag) => {
  const dispatch = useDispatch();
  const { userId } = getAuthenticatedUser();

  const [decision] = useDecision(flag, { autoUpdate: true }, { id: userId.toString() });
  const { active, variationKey } = decision || {};

  useEffect(() => {
    dispatch(setExperiment({ flag, active, variationKey }));
  }, [dispatch, flag, active, variationKey]);
};

export default useOptimizelyExperiment;
