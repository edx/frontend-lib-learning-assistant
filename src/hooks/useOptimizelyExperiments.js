import { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { OptimizelyDecideOption } from '@optimizely/react-sdk';
import { getAuthenticatedUser } from '@edx/frontend-platform/auth';
import { setExperiments } from '../data/slice';
import optimizely from '../data/optimizely'; // eslint-disable-line no-unused-vars

const useOptimizelyExperiments = async () => {
  const dispatch = useDispatch();
  const { userId } = getAuthenticatedUser();

  await optimizely.onReady();

  useEffect(() => {
    const user = optimizely.createUserContext(userId);
    const decisions = user.decideAll([OptimizelyDecideOption.ENABLED_FLAGS_ONLY]);

    dispatch(setExperiments({ decisions }));
  }, [dispatch, userId]);
};

export default useOptimizelyExperiments;
