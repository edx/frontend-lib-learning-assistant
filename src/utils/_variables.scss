$dark-green: #0e3639;
$accent-yellow: #f0cc00;
