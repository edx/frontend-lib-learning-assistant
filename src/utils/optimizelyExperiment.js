import { getOptimizely } from '../data/optimizely';

const trackChatBotMessageOptimizely = (userId, userAttributes = {}) => {
  const optimizelyInstance = getOptimizely();

  if (!optimizelyInstance || Object.keys(optimizelyInstance).length === 0) { return; }

  optimizelyInstance.onReady().then(() => {
    optimizelyInstance.track('learning_assistant_chat_message', userId, userAttributes);
  });
};

const trackUpgradeButtonClickedOptimizely = (userId, userAttributes = {}) => {
  const optimizelyInstance = getOptimizely();

  if (!optimizelyInstance || Object.keys(optimizelyInstance).length === 0) { return; }

  optimizelyInstance.onReady().then(() => {
    optimizelyInstance.track('upgrade_button_click', userId, userAttributes);
  });
};

export {
  trackChatBotMessageOptimizely,
  trackUpgradeButtonClickedOptimizely,
};
