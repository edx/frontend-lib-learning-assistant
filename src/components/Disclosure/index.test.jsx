import React from 'react';
import { screen } from '@testing-library/react';
import { render } from '../../utils/utils.test';
import { useCourseUpgrade } from '../../hooks';

import TrialDisclosure from '.';

const mockedUpgradeUrl = 'https://upgrade.edx/course/test';
const mockedAuditTrialDays = 7;

const PRIVACY_POLICY_URL = 'https://some.url/policy';
jest.mock('@edx/frontend-platform/config', () => ({
  ensureConfig: jest.fn(),
  getConfig: () => ({ PRIVACY_POLICY_URL }),
}));

jest.mock('../../hooks', () => ({
  useCourseUpgrade: jest.fn(),
  useTrackEvent: jest.fn(() => ({ track: jest.fn() })),
}));

describe('<TrialDisclosure />', () => {
  let container;

  describe('When trial upgrade is not being shown', () => {
    beforeEach(() => {
      useCourseUpgrade.mockReturnValue({ upgradeable: false });
      ({ container } = render(<TrialDisclosure><span>Children</span></TrialDisclosure>));
    });

    it('should have a link to the privacy policy url', () => {
      const link = screen.queryByText('privacy policy');

      expect(link).toBeInTheDocument();
      expect(link).toHaveAttribute('href', PRIVACY_POLICY_URL);
    });

    it('should not show the trial message', () => {
      const upgrade = screen.queryByTestId('free-days-label');

      expect(upgrade).not.toBeInTheDocument();
    });

    it('should not show the upgrade CTA', () => {
      const upgradeCta = screen.queryByTestId('upgrade-cta');

      expect(upgradeCta).not.toBeInTheDocument();
    });

    it('should match snapshot', () => {
      expect(container).toMatchSnapshot();
    });
  });

  describe('When trial upgrade being shown', () => {
    beforeEach(() => {
      useCourseUpgrade.mockReturnValue({
        upgradeable: true,
        upgradeUrl: mockedUpgradeUrl,
        auditTrialLengthDays: mockedAuditTrialDays,
      });
      ({ container } = render(<TrialDisclosure showTrial><span>Children</span></TrialDisclosure>));
    });

    it('should show the trial message', () => {
      const upgrade = screen.queryByTestId('free-days-label');

      expect(upgrade.textContent).toBe(`Free for ${mockedAuditTrialDays} days, then upgrade course for full access to Xpert features.`);
    });

    it('should show the trial button with the proper href to upgrade', () => {
      const upgradeCta = screen.queryByTestId('upgrade-cta');

      expect(upgradeCta).toBeInTheDocument();
      expect(upgradeCta).toHaveAttribute('href', mockedUpgradeUrl);
    });

    it('should match snapshot', () => {
      expect(container).toMatchSnapshot();
    });
  });
});
