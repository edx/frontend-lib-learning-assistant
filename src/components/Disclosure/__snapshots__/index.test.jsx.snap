// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrialDisclosure /> When trial upgrade being shown should match snapshot 1`] = `
<div>
  <section
    class="disclosure d-flex flex-column align-items-stretch"
  >
    <h2
      class="text-light-100"
    >
      Xpert Learning Assistant
    </h2>
    <div
      class="info"
    >
      <h3
        class="small py-2"
      >
        An AI-powered educational tool
      </h3>
      <div
        class="d-flex flex-column"
      >
        <div
          class="text-light-100 d-flex flex-row mb-3"
        >
          <span
            class="pgn__icon bullet-icon"
          >
            <svg
              aria-hidden="true"
              fill="none"
              focusable="false"
              height="24"
              role="img"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 17c-.83 0-1.5-.67-1.5-1.5h3c0 .83-.67 1.5-1.5 1.5zm3-2.5H9V15h6v1.5zm-.03-2.5H9.03A4.966 4.966 0 0 1 7 10c0-2.76 2.24-5 5-5s5 2.24 5 5c0 1.64-.8 3.09-2.03 4z"
                fill="currentColor"
              />
            </svg>
          </span>
          <div>
            Understand a concept
            <br />
            <small>
              “How does photosynthesis work?”
            </small>
          </div>
        </div>
        <div
          class="text-light-100 d-flex flex-row mb-4"
        >
          <span
            class="pgn__icon bullet-icon"
          >
            <svg
              aria-hidden="true"
              fill="none"
              focusable="false"
              height="24"
              role="img"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15 4v7H5.17l-.59.59-.58.58V4h11Zm1-2H3c-.55 0-1 .45-1 1v14l4-4h10c.55 0 1-.45 1-1V3c0-.55-.45-1-1-1Zm5 4h-2v9H6v2c0 .55.45 1 1 1h11l4 4V7c0-.55-.45-1-1-1Z"
                fill="currentColor"
              />
            </svg>
          </span>
          <div>
            Summarize your learning
            <br />
            <small>
              “Can you help me review pivot tables?”
            </small>
          </div>
        </div>
      </div>
      <div
        class="trial-period"
      >
        <div
          class="trial-period-content"
        >
          <div
            class="d-flex flex-row text-light-100"
          >
            <span
              class="pgn__icon bullet-icon bullet-icon"
            >
              <svg
                aria-hidden="true"
                fill="none"
                focusable="false"
                height="24"
                role="img"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m19 9 1.25-2.75L23 5l-2.75-1.25L19 1l-1.25 2.75L15 5l2.75 1.25L19 9zm-7.5.5L9 4 6.5 9.5 1 12l5.5 2.5L9 20l2.5-5.5L17 12l-5.5-2.5zM19 15l-1.25 2.75L15 19l2.75 1.25L19 23l1.25-2.75L23 19l-2.75-1.25L19 15z"
                  fill="currentColor"
                />
              </svg>
            </span>
            <small
              data-testid="free-days-label"
            >
              Free for 
              7 days
              , then upgrade course for full access to Xpert features.
            </small>
          </div>
          <a
            class="trial-upgrade mt-3 btn btn-brand btn-block"
            data-testid="upgrade-cta"
            href="https://upgrade.edx/course/test"
            target="_blank"
          >
            Upgrade now
          </a>
        </div>
      </div>
      <p
        class="disclaimer small text-light-100 py-3"
      >
        Note: This chat is AI generated, mistakes are possible. By using it you agree that edX may create a record of this chat. Your personal data will be used as described in our  
        <a
          class="pgn__hyperlink default-link standalone-link privacy-policy-link text-light-100"
          href="https://some.url/policy"
          rel="noopener noreferrer"
          target="_blank"
        >
          privacy policy
          <span
            class="pgn__hyperlink__external-icon"
            title="Opens in a new tab"
          >
            <span
              class="pgn__icon"
              data-testid="hyperlink-icon"
              style="height: 1em; width: 1em;"
            >
              <svg
                aria-hidden="true"
                fill="none"
                focusable="false"
                height="24"
                role="img"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 19H5V5h7V3H3v18h18v-9h-2v7ZM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7Z"
                  fill="currentColor"
                />
              </svg>
              <span
                class="sr-only"
              >
                in a new tab
              </span>
            </span>
          </span>
        </a>
        .
      </p>
    </div>
    <footer>
      <span>
        Children
      </span>
    </footer>
  </section>
</div>
`;

exports[`<TrialDisclosure /> When trial upgrade is not being shown should match snapshot 1`] = `
<div>
  <section
    class="disclosure d-flex flex-column align-items-stretch"
  >
    <h2
      class="text-light-100"
    >
      Xpert Learning Assistant
    </h2>
    <div
      class="info"
    >
      <h3
        class="small py-2"
      >
        An AI-powered educational tool
      </h3>
      <div
        class="d-flex flex-column"
      >
        <div
          class="text-light-100 d-flex flex-row mb-3"
        >
          <span
            class="pgn__icon bullet-icon"
          >
            <svg
              aria-hidden="true"
              fill="none"
              focusable="false"
              height="24"
              role="img"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 17c-.83 0-1.5-.67-1.5-1.5h3c0 .83-.67 1.5-1.5 1.5zm3-2.5H9V15h6v1.5zm-.03-2.5H9.03A4.966 4.966 0 0 1 7 10c0-2.76 2.24-5 5-5s5 2.24 5 5c0 1.64-.8 3.09-2.03 4z"
                fill="currentColor"
              />
            </svg>
          </span>
          <div>
            Understand a concept
            <br />
            <small>
              “How does photosynthesis work?”
            </small>
          </div>
        </div>
        <div
          class="text-light-100 d-flex flex-row mb-4"
        >
          <span
            class="pgn__icon bullet-icon"
          >
            <svg
              aria-hidden="true"
              fill="none"
              focusable="false"
              height="24"
              role="img"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15 4v7H5.17l-.59.59-.58.58V4h11Zm1-2H3c-.55 0-1 .45-1 1v14l4-4h10c.55 0 1-.45 1-1V3c0-.55-.45-1-1-1Zm5 4h-2v9H6v2c0 .55.45 1 1 1h11l4 4V7c0-.55-.45-1-1-1Z"
                fill="currentColor"
              />
            </svg>
          </span>
          <div>
            Summarize your learning
            <br />
            <small>
              “Can you help me review pivot tables?”
            </small>
          </div>
        </div>
      </div>
      <p
        class="disclaimer small text-light-100 py-3"
      >
        Note: This chat is AI generated, mistakes are possible. By using it you agree that edX may create a record of this chat. Your personal data will be used as described in our  
        <a
          class="pgn__hyperlink default-link standalone-link privacy-policy-link text-light-100"
          href="https://some.url/policy"
          rel="noopener noreferrer"
          target="_blank"
        >
          privacy policy
          <span
            class="pgn__hyperlink__external-icon"
            title="Opens in a new tab"
          >
            <span
              class="pgn__icon"
              data-testid="hyperlink-icon"
              style="height: 1em; width: 1em;"
            >
              <svg
                aria-hidden="true"
                fill="none"
                focusable="false"
                height="24"
                role="img"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 19H5V5h7V3H3v18h18v-9h-2v7ZM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7Z"
                  fill="currentColor"
                />
              </svg>
              <span
                class="sr-only"
              >
                in a new tab
              </span>
            </span>
          </span>
        </a>
        .
      </p>
    </div>
    <footer>
      <span>
        Children
      </span>
    </footer>
  </section>
</div>
`;
